File: ios/KampanyaRadari/KampanyaRadari/ViewModels/CampaignListViewModel.swift
import Foundation
import Observation

enum CampaignSortOption: String, CaseIterable, Identifiable {
    case expiringSoon = "Yakinda biten"
    case bestScore = "Firsat skoru"
    case bank = "Banka adi"
    case title = "Kampanya adi"

    var id: String { rawValue }
}

struct CategorySummary: Identifiable, Hashable {
    let name: String
    let count: Int

    var id: String { name }
}

enum CampaignCategory: String, CaseIterable, Identifiable {
    case fuel = "Akaryakıt"
    case electronics = "Elektronik"
    case fashion = "Giyim"
    case market = "Market"
    case online = "Online"
    case restaurant = "Restoran"
    case travel = "Seyahat"

    var id: String { rawValue }

    var iconName: String {
        switch self {
        case .fuel: return "fuelpump.fill"
        case .electronics: return "desktopcomputer"
        case .fashion: return "tshirt.fill"
        case .market: return "cart.fill"
        case .online: return "network"
        case .restaurant: return "fork.knife"
        case .travel: return "airplane"
        }
    }
}

@MainActor
@Observable
final class CampaignListViewModel {
    var campaigns: [Campaign] = []
    private var categoryCache: [String: String] = [:]
    var searchText = ""
    var selectedBanks: Set<String> = []
    var selectedCategories: Set<String> = []
    var selectedRewardType: String?
    var showFavoritesOnly = false
    var showMyCardsOnly = false
    var sortOption: CampaignSortOption = .expiringSoon
    var isLoading = false
    var errorMessage: String?
    var campaignsRevision = 0

    private let service = CampaignService()

    var banks: [String] {
        Array(Set(campaigns.map(\.bank))).sorted {
            label(for: $0).localizedCaseInsensitiveCompare(label(for: $1)) == .orderedAscending
        }
    }

    var categories: [String] {
        CampaignCategory.allCases
            .map(\.rawValue)
            .filter { category in
                campaigns.contains { canonicalCategory(for: $0) == category }
            }
    }

    var rewardTypes: [String] {
        sortedUniqueValues(campaigns.compactMap(\.rewardType))
    }

    var categorySummaries: [CategorySummary] {
        CampaignCategory.allCases.compactMap { category in
            let count = campaigns.filter { canonicalCategory(for: $0) == category.rawValue }.count
            guard count > 0 else { return nil }
            return CategorySummary(name: category.rawValue, count: count)
        }
    }

    var topCategorySummaries: [CategorySummary] {
        categorySummaries
    }

    var bankCount: Int {
        Set(campaigns.map(\.bank)).count
    }

    var hasAdvancedFilters: Bool {
        !selectedBanks.isEmpty || !selectedCategories.isEmpty || selectedRewardType != nil || showFavoritesOnly || showMyCardsOnly || sortOption != .expiringSoon
    }

    var hasContentFilters: Bool {
        !selectedCategories.isEmpty || selectedRewardType != nil || showFavoritesOnly || showMyCardsOnly
    }

    func filteredCampaigns(favoriteIDs: Set<String>, myCardBanks: Set<String>) -> [Campaign] {
        let filtered = campaigns.filter { campaign in
            let bankMatches = selectedBanks.isEmpty || selectedBanks.contains(campaign.bank)
            let categoryMatches = selectedCategories.isEmpty || selectedCategories.contains(canonicalCategory(for: campaign))
            let rewardMatches = selectedRewardType == nil || campaign.rewardType == selectedRewardType
            let favoriteMatches = !showFavoritesOnly || favoriteIDs.contains(campaign.id)
            let myCardsMatch = !showMyCardsOnly || myCardBanks.contains(campaign.bank)
            let search = searchText.trimmingCharacters(in: .whitespacesAndNewlines)
            let searchMatches = search.isEmpty
                || campaign.title.localizedCaseInsensitiveContains(search)
                || campaign.displaySummary.localizedCaseInsensitiveContains(search)
                || campaign.bank.localizedCaseInsensitiveContains(search)

            return bankMatches && categoryMatches && rewardMatches && favoriteMatches && myCardsMatch && searchMatches
        }

        return sort(filtered)
    }

    func resetFilters() {
        selectedBanks = []
        selectedCategories = []
        selectedRewardType = nil
        showFavoritesOnly = false
        showMyCardsOnly = false
        sortOption = .expiringSoon
    }

    func clearAdvancedFilters() {
        selectedCategories = []
        selectedRewardType = nil
        showFavoritesOnly = false
        showMyCardsOnly = false
    }

    func showAllCampaigns() {
        resetFilters()
        searchText = ""
    }

    func showCampaigns(category: String) {
        searchText = ""
        selectedBanks = []
        selectedCategories = [category]
        selectedRewardType = nil
        showFavoritesOnly = false
        showMyCardsOnly = false
        sortOption = .expiringSoon
    }

    func showFavoriteCampaigns() {
        searchText = ""
        selectedBanks = []
        selectedCategories = []
        selectedRewardType = nil
        showFavoritesOnly = true
        showMyCardsOnly = false
        sortOption = .expiringSoon
    }

    func showMyCardCampaigns() {
        searchText = ""
        selectedBanks = []
        selectedCategories = []
        selectedRewardType = nil
        showFavoritesOnly = false
        showMyCardsOnly = true
        sortOption = .expiringSoon
    }

    func toggleBank(_ bank: String) {
        if selectedBanks.contains(bank) {
            selectedBanks.remove(bank)
        } else {
            selectedBanks.insert(bank)
        }
    }

    func clearBanks() {
        selectedBanks = []
    }

    func toggleCategory(_ category: String) {
        if selectedCategories.contains(category) {
            selectedCategories.remove(category)
        } else {
            selectedCategories.insert(category)
        }
    }

    func clearCategories() {
        selectedCategories = []
    }

    func label(for bank: String) -> String {
        campaigns.first(where: { $0.bank == bank })?.displayBank ?? bank
    }

    func iconName(for category: String) -> String {
        CampaignCategory.allCases.first(where: { $0.rawValue == category })?.iconName ?? "tag.fill"
    }

    private func sortedUniqueValues(_ values: [String]) -> [String] {
        Array(Set(values.filter { !$0.isEmpty })).sorted {
            $0.localizedCaseInsensitiveCompare($1) == .orderedAscending
        }
    }

    private func canonicalCategory(for campaign: Campaign) -> String {
        if let cached = categoryCache[campaign.id] { return cached }
        let result = Self.computeCanonicalCategory(for: campaign)
        categoryCache[campaign.id] = result
        return result
    }

    nonisolated static func computeCanonicalCategory(for campaign: Campaign) -> String {
        let haystack = [
            campaign.title,
            campaign.summary,
            campaign.description
        ]
            .compactMap { $0 }
            .joined(separator: " ")
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "tr_TR"))
            .lowercased()

        if haystack.contains("akaryakit") || haystack.contains("yakit") || haystack.contains("petrol") || haystack.contains("shell") || haystack.contains("opet") || haystack.contains("total") {
            return CampaignCategory.fuel.rawValue
        }
        if haystack.contains("elektronik") || haystack.contains("teknosa") || haystack.contains("mediamarkt") || haystack.contains("vatan") || haystack.contains("bilgisayar") || haystack.contains("telefon") {
            return CampaignCategory.electronics.rawValue
        }
        if haystack.contains("giyim") || haystack.contains("moda") || haystack.contains("zara") || haystack.contains("defacto") || haystack.contains("lc waikiki") || haystack.contains("ipekyol") || haystack.contains("vakko") {
            return CampaignCategory.fashion.rawValue
        }
        if haystack.contains("market") || haystack.contains("migros") || haystack.contains("a101") || haystack.contains("sok") || haystack.contains("carrefour") {
            return CampaignCategory.market.rawValue
        }
        if haystack.contains("restoran") || haystack.contains("restaurant") || haystack.contains("cafe") || haystack.contains("yemek") || haystack.contains("bigchefs") || haystack.contains("cookshop") {
            return CampaignCategory.restaurant.rawValue
        }
        if haystack.contains("seyahat")
            || haystack.contains("otel")
            || haystack.contains("hotel")
            || haystack.contains("konaklama")
            || haystack.contains("ucak")
            || haystack.contains("ucak bileti")
            || haystack.contains("otobus bileti")
            || haystack.contains("obilet")
            || haystack.contains("tatil")
            || haystack.contains("havalimani")
            || haystack.contains("lounge")
            || haystack.contains("transfer")
            || haystack.contains("arac kiralama")
            || haystack.contains("yurt disi cikis")
            || haystack.contains("cikis harci")
            || haystack.contains("pazaramatatil")
            || haystack.contains("tatilbudur")
            || haystack.contains("jolly")
            || haystack.contains("yolcu360")
            || haystack.contains("enuygun") {
            return CampaignCategory.travel.rawValue
        }
        return CampaignCategory.online.rawValue
    }

    private func sort(_ campaigns: [Campaign]) -> [Campaign] {
        switch sortOption {
        case .expiringSoon:
            return campaigns.sorted {
                ($0.validTo ?? .distantFuture, $0.title) < ($1.validTo ?? .distantFuture, $1.title)
            }
        case .bestScore:
            return campaigns.sorted {
                ($0.opportunityScore ?? 0, $0.title) > ($1.opportunityScore ?? 0, $1.title)
            }
        case .bank:
            return campaigns.sorted {
                let bankOrder = $0.displayBank.localizedCaseInsensitiveCompare($1.displayBank)
                if bankOrder != .orderedSame { return bankOrder == .orderedAscending }
                return $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending
            }
        case .title:
            return campaigns.sorted {
                $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending
            }
        }
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            let fetched = try await service.fetchActiveCampaigns()
            let cache: [String: String] = await Task.detached(priority: .userInitiated) {
                Dictionary(
                    uniqueKeysWithValues: fetched.map {
                        ($0.id, CampaignListViewModel.computeCanonicalCategory(for: $0))
                    }
                )
            }.value
            categoryCache = cache
            campaigns = fetched
            campaignsRevision += 1
        } catch {
            errorMessage = "Kampanyalar yuklenemedi. \(error.localizedDescription)"
        }
        isLoading = false
    }
}

File: android/app/src/main/java/com/mericucan/kampanyaradari/viewmodel/CampaignListViewModel.kt
package com.mericucan.kampanyaradari.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mericucan.kampanyaradari.data.model.Campaign
import com.mericucan.kampanyaradari.data.remote.CampaignService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate

// ── Enums (iOS'tan birebir) ───────────────────────────────────

enum class SortOption(val label: String) {
    EXPIRING_SOON("Yakında biten"),
    BEST_SCORE("Fırsat skoru"),
    BANK("Banka adı"),
    TITLE("Kampanya adı")
}

enum class CampaignCategory(val label: String, val emoji: String) {
    FUEL("Akaryakıt", "⛽"),
    ELECTRONICS("Elektronik", "💻"),
    FASHION("Giyim", "👕"),
    MARKET("Market", "🛒"),
    ONLINE("Online", "🌐"),
    RESTAURANT("Restoran", "🍽️"),
    TRAVEL("Seyahat", "✈️")
}

data class CategorySummary(val category: CampaignCategory, val count: Int)

// ── ViewModel ─────────────────────────────────────────────────

class CampaignListViewModel : ViewModel() {
    private val service = CampaignService()
    private val categoryCache = mutableMapOf<String, String>()

    private val _campaigns    = MutableStateFlow<List<Campaign>>(emptyList())
    private val _isLoading    = MutableStateFlow(false)
    private val _lastSyncTime = MutableStateFlow<Long?>(null)
    private val _error       = MutableStateFlow<String?>(null)

    val campaigns: StateFlow<List<Campaign>>  = _campaigns.asStateFlow()
    val isLoading: StateFlow<Boolean>         = _isLoading.asStateFlow()
    val lastSyncTime: StateFlow<Long?>        = _lastSyncTime.asStateFlow()
    val error: StateFlow<String?>            = _error.asStateFlow()

    // Filter state
    val searchText          = MutableStateFlow("")
    val selectedBanks       = MutableStateFlow<Set<String>>(emptySet())
    val selectedCategories  = MutableStateFlow<Set<String>>(emptySet())
    val selectedRewardType  = MutableStateFlow<String?>(null)
    val showFavoritesOnly   = MutableStateFlow(false)
    val showMyCardsOnly     = MutableStateFlow(false)
    val sortOption          = MutableStateFlow(SortOption.EXPIRING_SOON)

    init { load() }

    fun load() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                val fetched = service.fetchActiveCampaigns()
                fetched.forEach { categoryCache[it.id] = computeCategory(it) }
                _campaigns.value = fetched
                _lastSyncTime.value = System.currentTimeMillis()
            } catch (e: Exception) {
                _error.value = "Kampanyalar yüklenemedi: ${e.message}"
            }
            _isLoading.value = false
        }
    }

    fun filteredCampaigns(favoriteIds: Set<String>, myCardBanks: Set<String>): List<Campaign> {
        val search    = searchText.value.trim()
        val banks     = selectedBanks.value
        val cats      = selectedCategories.value
        val reward    = selectedRewardType.value
        val favOnly   = showFavoritesOnly.value
        val cardsOnly = showMyCardsOnly.value

        return _campaigns.value
            .filter { c ->
                (banks.isEmpty() || banks.contains(c.bank)) &&
                (cats.isEmpty()  || cats.contains(getCategory(c))) &&
                (reward == null  || c.rewardType?.contains(reward, ignoreCase = true) == true) &&
                (!favOnly        || favoriteIds.contains(c.id)) &&
                (!cardsOnly      || myCardBanks.contains(c.bank)) &&
                (search.isEmpty() ||
                    c.title.contains(search, ignoreCase = true) ||
                    c.displaySummary.contains(search, ignoreCase = true) ||
                    c.bank.contains(search, ignoreCase = true))
            }
            .let { sorted(it) }
    }

    fun getCategory(campaign: Campaign): String =
        categoryCache.getOrPut(campaign.id) { computeCategory(campaign) }

    fun allBanks(): List<String> =
        _campaigns.value.map { it.bank }.toSet().sortedWith(compareBy { bankLabel(it) })

    fun bankLabel(bank: String): String =
        _campaigns.value.firstOrNull { it.bank == bank }?.displayBank ?: bank

    fun categorySummaries(): List<CategorySummary> =
        CampaignCategory.entries.mapNotNull { cat ->
            val count = _campaigns.value.count { getCategory(it) == cat.label }
            if (count > 0) CategorySummary(cat, count) else null
        }

    fun hasActiveFilters(): Boolean =
        selectedBanks.value.isNotEmpty() ||
        selectedCategories.value.isNotEmpty() ||
        selectedRewardType.value != null ||
        showFavoritesOnly.value ||
        showMyCardsOnly.value ||
        sortOption.value != SortOption.EXPIRING_SOON

    fun resetFilters() {
        selectedBanks.value      = emptySet()
        selectedCategories.value = emptySet()
        selectedRewardType.value = null
        showFavoritesOnly.value  = false
        showMyCardsOnly.value    = false
        sortOption.value         = SortOption.EXPIRING_SOON
    }

    fun showFavoriteCampaigns() {
        resetFilters()
        showFavoritesOnly.value = true
    }

    fun showMyCardCampaigns() {
        resetFilters()
        showMyCardsOnly.value = true
    }

    fun showCampaigns(category: String) {
        resetFilters()
        selectedCategories.value = setOf(category)
    }

    // ── Private helpers ───────────────────────────────────────

    private fun sorted(list: List<Campaign>): List<Campaign> = when (sortOption.value) {
        SortOption.EXPIRING_SOON -> list.sortedWith(
            compareBy({ it.validToDate ?: LocalDate.MAX }, { it.title })
        )
        SortOption.BEST_SCORE -> list.sortedWith(
            compareByDescending<Campaign> { it.opportunityScore ?: 0 }.thenBy { it.title }
        )
        SortOption.BANK  -> list.sortedWith(compareBy({ it.displayBank }, { it.title }))
        SortOption.TITLE -> list.sortedBy { it.title }
    }

    private fun computeCategory(c: Campaign): String {
        val h = listOfNotNull(c.title, c.summary, c.description)
            .joinToString(" ")
            .lowercase()
            .replace('ı', 'i').replace('ş', 's').replace('ğ', 'g')
            .replace('ü', 'u').replace('ö', 'o').replace('ç', 'c')

        return when {
            h.any("akaryakit", "yakit", "petrol", "shell", "opet", "total") -> CampaignCategory.FUEL.label
            h.any("elektronik", "teknosa", "mediamarkt", "vatan", "bilgisayar", "telefon") -> CampaignCategory.ELECTRONICS.label
            h.any("giyim", "moda", "zara", "defacto", "lc waikiki", "ipekyol", "vakko") -> CampaignCategory.FASHION.label
            h.any("market", "migros", "a101", "sok", "carrefour", "bim", "file", "metro", "kipa", "hakmar", "makro") -> CampaignCategory.MARKET.label
            h.any("restoran", "restaurant", "cafe", "yemek", "bigchefs") -> CampaignCategory.RESTAURANT.label
            h.any("seyahat", "otel", "hotel", "ucak", "tatil", "havalimani", "lounge", "transfer",
                "arac kiralama", "tatilbudur", "jolly", "yolcu360", "enuygun") -> CampaignCategory.TRAVEL.label
            else -> CampaignCategory.ONLINE.label
        }
    }

    private fun String.any(vararg terms: String) = terms.any { this.contains(it) }
}
